predicators/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -39,12 +40,159 @@ def _create_general_pnad_for_option(
+                                        add_effects=updated_add_effects)
+        new_pnad = PartialNSRTAndDatastore(new_pnad_op, [], pnad.option_spec)
+        # Note: we don't need to copy anything related to keep effects into
+        # new_pnad here, because we only care about keep effects on the final
+        # iteration of backchaining, where this function is never called.
+
+        return new_pnad
+
+    @staticmethod
+    def get_pnads_with_keep_effects(
+            pnad: PartialNSRTAndDatastore) -> Set[PartialNSRTAndDatastore]:
+        """Return a new set of PNADs that include keep effects into the given
+        PNAD."""
+        # The keep effects that we want are the subset of possible keep
+        # effects which are not already in the PNAD's add effects, and
+        # whose predicates were either (i) determined to be ignore effects,
+        # or (ii) in the delete effects.
+        keep_effects = {
+            eff
+            for eff in pnad.poss_keep_effects
+            if eff not in pnad.op.add_effects and (
+                eff.predicate in pnad.op.ignore_effects
+                or eff in pnad.op.delete_effects)
+        }
+        new_pnads_with_keep_effects = set()
+        # Given these keep effects, we need to create a combinatorial number of
+        # PNADs, one for each unique combination of keep effects. Moreover, we
+        # need to ensure that they are named differently from each other. Some
+        # of these PNADs will be filtered out later if they are not useful to
+        # cover any datapoints.
+        for r in range(1, len(keep_effects) + 1):
+            for keep_effects_subset in itertools.combinations(keep_effects, r):
+                # These keep effects (keep_effects_subset) could involve new
+                # variables, which we need to add to the PNAD parameters.
+                params_set = set(pnad.op.parameters)
+                for eff in keep_effects_subset:
+                    for var in eff.variables:
+                        params_set.add(var)
+                parameters = sorted(params_set)
+                # The keep effects go into both the PNAD preconditions and the
+                # PNAD add effects.
+                preconditions = pnad.op.preconditions | set(
+                    keep_effects_subset)
+                add_effects = pnad.op.add_effects | set(keep_effects_subset)
+                # Create the new PNAD.
+                new_pnad_op = pnad.op.copy_with(parameters=parameters,
+                                                preconditions=preconditions,
+                                                add_effects=add_effects)
+                new_pnad = PartialNSRTAndDatastore(new_pnad_op, [],
+                                                   pnad.option_spec)
+                # Remember to copy seg_to_keep_effects_sub into the new_pnad!
+                new_pnad.seg_to_keep_effects_sub = pnad.seg_to_keep_effects_sub
```

Comment:
can we just do this in `copy_with()` above? if that's more than a 1 line change though dwbi

Reply:
`copy_with` is only defined for `pnad.op` I think.